docs/proposals/connector-protocol-v2-addendum-migration.md

```diff
@@ -0,0 +1,127 @@
+
+
+## Protocol State Machine
+
+Each connection must follow the following state machine definition.
+
+```
+      +-------------------+----------------------------+
+      |                   |                            |
+ +-----------+       +-----------+     +-----------+   |
+ | Connected +------>| Handshake +---->| Streaming |---+
+ +-----------+       +-----------+     +-----------+   |
+  ^      |                |                  |         |
+  |      |                +--------+         |         |
+  |      |                V        |         |         V
+  |      |            +--------+   |         |   +-----------+
+  |      +----------> | Error  | <-----------+---| Migrating |
+  |      |            +--------+   |         |   +-----------+
+  |      |                |        |         |         |
+  |      |                V        V         |         |
+  |      |       +-------------------+       |         |
+  |      +------>+  Disconnected     |<------+---------+
+  |              +-------------------+                 |
+  |                                                    |
+  +----------------------------------------------------+
+```
+
+In the connected state the following frame types are valid:
+
+    - HELLO: Connector -> Worker (next state = handshake)
+    - ERROR: * -> * (next state = error)
+    - MIGRATE: Worker -> Connector
+
+In the handshake state the following frame types are valid:
+
+    - OK: Worker -> Connector (next state = streaming)
+    - ERROR: * -> * (next state = error)
+    - MIGRATE: Worker -> Connector (next state = migrating)
+
+In the streaming state the following frame types are valid:
+
+   - NOFITY: Connector -> Worker
+   - NOTIFY_ACK: Worker -> Connector
+   - MESSAGE: Connector -> Worker
+   - ACK: Worker -> Connector
+   - RESTART: Worker -> Connector
+   - ERROR: * -> * (next state = error)
+   - MIGRATE: Worker -> Connector (next state = migrating)
+
+In the migrating state, the following frame types are valid:
+  - MIGRATE_OK: Connector -> Worker (next state = connected)
+  - ERROR: * -> * (next state = error)
+
+
+## Migration Protocol
+
+In order to accomplish a migration, the following sequence of event needs to occur:
+
+| order | Connector | Worker |  Notes |
+| - | - | - | - |
+| 0 | Connect | - | Connector is in one of {Connected, Handshake, Steaming} |
+| 1 | - | send MIGRATE | This message contains information about the new target address |
+| 2 | close all streams | - | - |
+| 3 | send EOS for all streams | - | - |
+| 4 | - | checkpoint all streams | - |
+| 5 | - | migrate source state to new worker | - |
+| 6 | - | Ack all streams | - |
+| 7 | connect to new target address | - | Old connection is still open |
+| 8 | Handshake on new address | - | Old connection is still open |
+| 9 | send MIGRATION_ACK | - | Send this on old connection |
+| 10 | - | Worker closes connection, completes shut down | - |
+| 11 | resume streaming | - | on new connection; Send NOTIFYs, etc. | - |
```

Comment:
I have some concerns around the potential waiting that can happen on steps 4-10. There are a lot of points where we need to be able to deal with timeouts and failures of either party in the protocol. Also, there is an issue of being in a migrating state when receiving a new migration request before the other migration is complete.

As noted below, the error states after failure on the connection or worker side should leave the connector with enough information to direct connection attempts to the new worker. Timeouts handle this and the handshake and other protocol features scrub inconsistent progress. This is the key guarantee. Everything else is about making this more graceful and efficient.

However, I don't think it's quite enough yet to make the guarantee in the case where things take awhile and possibly some impatient part of the system decides to change the migration plan. It's important that migration requests be sent down the latest active session for a connector. Once a worker sends a migration request, the channel should also be in a third state: flushing streams (or whatever we want to call waiting for it to drain).

The new connection can go through the handshake and migration but now we also have a race condition for the same error. If a migration is needed again at some point between the first and the second successful connection, then we have an issue and we'll need to use state to check every connection and remigration could ping-pong between nodes if it's not perfectly synchronized.

There are a few fixes. Some involve restraining when migrations occur in the cluster as a cluster-wide settlement (so the cluster must settle before making new changes). This is reasonable as it keeps the state on each worker consistent through means we already use. It does present the same partition fragility and issues in very large cluster scaling but it's a practical trade.

An alternative is to use a migration id/token. I'm throwing the idea out here though it's not necessarily better, but it might help discussion:

This is a bit lighter than cluster-wide state (depending on the implementation of course). What we'd do here is ask for a migration and give an ID. The worker could stop giving credits with acks to enforce spin-down on that connection. Instead of coordinating directly on that connection, it's left open but not use for migration acks. Now we connect to a new machine and during the handshake, we send the migration frame instead of the hello frame, and pass along the token. This allows the new worker to then reach out to the old worker and ask for the state. This will have the old worker close the connection and flush it's buffers and ensure everything had an associated checkpoint. Finally, the migraiton_ack is sent to the new connection with a summary of stream state (I say summary since only the delta from the old worker's point of view from when it send the migration message and when it closed and awaited checkpointing is required). Now the handshake is done (new variant) and it can continue.

The handshake could fail with a new migration message saying that it's now out of date again and the new token repeats the process. Of course, having N connections open but migrating is not great if this happens rapidly. So instead it makes more sense to still have some safety around how rapidly migrations are allowed from a control interface perspective, but in theory it should be more resilient against misbehavior.